### PR TITLE
Java JNI (JPype) Wrapper

### PR DESCRIPTION
**Pull request by adriangonz**
**What this PR does / why we need it**:
Add an experimental option to wrap Java models which leverages the Python inference server under the hood. The communication between the Python and Java side is managed through JNI using JPype.

To make the interop more efficient, this PR also introduces a method to disable the payload serialisation / deserialisation on the Python side (so that is managed by Java). This is controlled through the `PAYLOAD_PASSTHROUGH` env variable. 

**To Do**:

- [x] Add some testing for JNI wrapper
- [x] Publish `0.3.0` version of Java wrapper
- [ ] ~Add example~ (moved to #2458)

**Which issue(s) this PR fixes**:

Work towards #1344

**Special notes for your reviewer**:
This works is considered as an incubating project.

**Does this PR introduce a user-facing change?**:

```release-note
Added incubating approach to wrap Java models, leveraging the Python inference server under the hood.
```

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **adriangonz**
You can assign the PR to them by writing `/assign @adriangonz` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Thu Sep  3 13:27:26 UTC 2020
The logs for [lint] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=2

**Comment by seldondev** [comment]
Thu Sep  3 13:27:32 UTC 2020
The logs for [pr-build] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=1

**Comment by adriangonz** [comment]
While this is still a WIP PR, it would be great to hear your thoughts about the changes introduced in this PR @cliveseldon @axsaucedo @RafalSkolasinski .

File: incubating/wrappers/s2i/java-jni/test/model-template-app/pom.xml
Comment (axsaucedo) [inline incubating/wrappers/s2i/java-jni/test/model-template-app/pom.xml:70]: Will this have to stay when the PR is merged?

File: python/seldon_core/wrapper.py
Comment (axsaucedo) [inline python/seldon_core/wrapper.py:26]: Would be good to document this in the python section and in the upgrading page

File: incubating/wrappers/s2i/java-jni/version.txt
Comment (axsaucedo) [inline incubating/wrappers/s2i/java-jni/version.txt:1]: If the version of the image is 1.2.3-dev, it normaly implies that it's built on the same "consistency" as all the other images, so for this it would be required to add to the build_and_push_all_images script in the top level repo, or potentially move it to a lower version.

Comment (axsaucedo) [reply incubating/wrappers/s2i/java-jni/version.txt:1]: In theory we've moved the components as they are graduated (eg the java wrapper isn't in verison .1.2.3, so not sure whether we'll want to keep it on same version or not. I guess my question would also be, what would be required to graduate? Maybe we can flesh out some of those - I guess if it's more on the stablity front instead of the feature side it could go on an almost pre-release

File: doc/source/java-jni/README.md
Comment (axsaucedo) [inline doc/source/java-jni/README.md:54]: I was going to ask if metrics and request tags are collected, but i realised they are, as metrics are collected right after predict_Raw, and tags are passed on by whatever is returned. It may be worth documenting this briefly in this section, to just mention that both are actually available

**Review by axsaucedo — COMMENTED** [review]
This is great @adriangonz ! I added a couple of comments, would be great to hear your thoughts

**Comment by seldondev** [comment]
Tue Sep  8 08:39:21 UTC 2020
The logs for [lint] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=4

**Comment by seldondev** [comment]
Tue Sep  8 08:39:20 UTC 2020
The logs for [pr-build] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=3

File: incubating/wrappers/s2i/java-jni/test/model-template-app/pom.xml
Comment (adriangonz) [reply incubating/wrappers/s2i/java-jni/test/model-template-app/pom.xml:70]: Good spot! We should be good to publish `0.3.0` of the `seldon-core-wrapper` package. It's still backwards compatible (i.e. it's still got Spring, Tomcat and everything else embedded).

File: python/seldon_core/wrapper.py
Comment (adriangonz) [reply python/seldon_core/wrapper.py:26]: By default that flag will be disabled, so the behaviour will be the same as it is currently. We can add a note on the docs though! I can see how it could be useful in some use cases outside of this PR.

File: incubating/wrappers/s2i/java-jni/version.txt
Comment (adriangonz) [reply incubating/wrappers/s2i/java-jni/version.txt:1]: I totally agree. I was using version `0.3.0` everywhere else, so I'm not sure how that `version.txt` leaked there. We should keep `0.3.0` as the version for now until it graduates.

File: doc/source/java-jni/README.md
Comment (adriangonz) [reply doc/source/java-jni/README.md:54]: Actually, they may not be available. We currently get metrics and tags from either the response payload or the `metrics()` method. However, we don't decode the payload at all on the Python side and we haven't hooked the `metrics()` method to the Java side yet. 

Default metrics coming from Flask should be available though.

**Comment by seldondev** [comment]
Mon Sep 14 12:05:34 UTC 2020
The logs for [lint] [6] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/6.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=6

**Comment by seldondev** [comment]
Mon Sep 14 12:05:34 UTC 2020
The logs for [pr-build] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=5

**Comment by seldondev** [comment]
Mon Sep 14 15:44:19 UTC 2020
The logs for [lint] [8] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/8.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=8

**Comment by seldondev** [comment]
Mon Sep 14 15:44:27 UTC 2020
The logs for [pr-build] [7] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/7.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=7

**Comment by seldondev** [comment]
Fri Sep 18 17:10:16 UTC 2020
The logs for [pr-build] [9] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/9.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=9

**Comment by seldondev** [comment]
Fri Sep 18 17:10:23 UTC 2020
The logs for [lint] [10] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/10.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=10

**Comment by adriangonz** [comment]
/cc @RafalSkolasinski @axsaucedo 

I just added some tests to this PR using the new JNI wrapper. I tried to also add some for the pure Java wrapper but there was some issue there. This can be tackled on a separate PR though.

File: doc/source/java-jni/README.md
Comment (axsaucedo) [reply doc/source/java-jni/README.md:54]: Interesting, ok this is good to know, may be worth making a note somewhere

File: incubating/wrappers/s2i/java-jni/java-jni/JavaJNIServer.py
Comment (axsaucedo) [inline incubating/wrappers/s2i/java-jni/java-jni/JavaJNIServer.py:28]: Better to move to github issue than todo in code

**Review by axsaucedo — APPROVED** [review]
THis is great! Nice one @adriangonz - should be good to go once integration passes (excluding known tests with issues)
/approve

File: incubating/wrappers/s2i/java-jni/java-jni/JavaJNIServer.py
Comment (adriangonz) [reply incubating/wrappers/s2i/java-jni/java-jni/JavaJNIServer.py:28]: Good spot! That one isn't actually needed anymore.

**Comment by seldondev** [comment]
New changes are detected. LGTM label has been removed.

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[axsaucedo](https://github.com/SeldonIO/seldon-core/pull/2374#pullrequestreview-491652595 "Approved")*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [axsaucedo]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Mon Sep 21 07:53:08 UTC 2020
The logs for [pr-build] [11] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/11.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=11

**Comment by seldondev** [comment]
Mon Sep 21 07:53:17 UTC 2020
The logs for [lint] [12] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2374/12.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2374 --build=12

**Pull request merged**
